aws/rust-runtime/aws-config/src/imds/client.rs

```diff
@@ -0,0 +1,893 @@
+/// variable. Valid values: `IPv4`, `IPv6`
+///
+/// 6. An [endpoint mode](EndpointMode) loaded from the `ec2_metadata_service_endpoint_mode` field in
+/// `~/.aws/config`:
+/// ```ini
+/// [default]
+/// # ... other configuration
+/// ec2_metadata_service_endpoint_mode = IPv4
+/// ```
+///
+/// 7. The default value of `http://169.254.169.254` will be used.
+///
+///
+pub struct Client {
+    endpoint: Endpoint,
+    inner: smithy_client::Client<DynConnector, ImdsMiddleware>,
+}
+
+impl Client {
+    /// IMDS client builder
+    pub fn builder() -> Builder {
+        Builder::default()
+    }
+
+    /// Retrieve information from IMDS
+    ///
+    /// This method will handle loading and caching a session token, combining the `path` with the
+    /// configured IMDS endpoint, and retrying potential errors.
+    ///
+    /// For more information about IMDSv2 methods and functionality, see
+    /// [Instance metadata and user data](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html)
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use aws_config::imds::client::Client;
+    /// # async fn docs() {
+    /// let client = Client::builder().build().await.expect("valid client");
+    /// let ami_id = client
+    ///   .get("/latest/meta-data/ami-id")
+    ///   .await
+    ///   .expect("failure communicating with IMDS");
+    /// # }
+    /// ```
+    pub async fn get(&self, path: &str) -> Result<String, SdkError<ImdsError>> {
+        let operation = self
+            .make_operation(path)
+            .map_err(|err| SdkError::ConstructionFailure(err.into()))?;
+        self.inner.call(operation).await
+    }
+
+    /// Creates a smithy_http Operation to for `path`
+    /// - Convert the path to a URI
+    /// - Set the base endpoint on the URI
+    /// - Add a user agent
+    fn make_operation(
+        &self,
+        path: &str,
+    ) -> Result<Operation<ImdsGetResponseHandler, ImdsErrorPolicy>, ImdsError> {
+        let mut base_uri: Uri = path.parse().map_err(|_| ImdsError::InvalidPath)?;
+        self.endpoint.set_endpoint(&mut base_uri, None);
+        let request = http::Request::builder()
+            .uri(base_uri)
+            .body(SdkBody::empty())
+            .expect("valid request");
+        let mut request = operation::Request::new(request);
+        request.properties_mut().insert(USER_AGENT);
+        Ok(Operation::new(request, ImdsGetResponseHandler)
+            .with_metadata(Metadata::new("get", "imds"))
+            .with_retry_policy(ImdsErrorPolicy))
+    }
+}
+
+/// An error retrieving metdata from IMDS
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum ImdsError {
+    /// An IMDSv2 Token could not be loaded
+    ///
+    /// Requests to IMDS must be accompanied by a token obtained via a `PUT` request. This is handled
+    /// transparently by the [`Client`].
+    FailedToLoadToken(SdkError<TokenError>),
```

Comment:
since the error comes out of smithy-client it's broken down by phase and nested inside of `SdkError` (eg. error constructing the request etc.)

I could embed the SDK error inside the token error, but since we actually probably want the raw response for debugging, it seemed simplest to just leave it as an SdkError